Index: linux-3.19.8/drivers/net/ethernet/intel/e1000/e1000_hw.c
===================================================================
--- linux-3.19.8.orig/drivers/net/ethernet/intel/e1000/e1000_hw.c	2015-05-11 14:34:10.000000000 +0200
+++ linux-3.19.8/drivers/net/ethernet/intel/e1000/e1000_hw.c	2015-06-08 14:15:44.905691368 +0200
@@ -4303,7 +4303,7 @@
 	/* Setup the receive address. */
 	e_dbg("Programming MAC Address into RAR[0]\n");
 
-	e1000_rar_set(hw, hw->mac_addr, 0);
+	e1000_rar_set(hw, hw->perm_mac_addr, 0);
 
 	rar_num = E1000_RAR_ENTRIES;
 
Index: linux-3.19.8/drivers/net/ethernet/intel/e1000/e1000_main.c
===================================================================
--- linux-3.19.8.orig/drivers/net/ethernet/intel/e1000/e1000_main.c	2015-05-11 14:34:10.000000000 +0200
+++ linux-3.19.8/drivers/net/ethernet/intel/e1000/e1000_main.c	2015-06-08 14:15:45.015690065 +0200
@@ -2221,7 +2221,7 @@
 	memcpy(netdev->dev_addr, addr->sa_data, netdev->addr_len);
 	memcpy(hw->mac_addr, addr->sa_data, netdev->addr_len);
 
-	e1000_rar_set(hw, hw->mac_addr, 0);
+	e1000_rar_set(hw, hw->mac_addr, E1000_RAR_ENTRIES - 2);
 
 	if (hw->mac_type == e1000_82542_rev2_0)
 		e1000_leave_82542_rst(adapter);
Index: linux-3.19.8/drivers/net/ethernet/intel/e1000e/mac.c
===================================================================
--- linux-3.19.8.orig/drivers/net/ethernet/intel/e1000e/mac.c	2015-05-11 14:34:10.000000000 +0200
+++ linux-3.19.8/drivers/net/ethernet/intel/e1000e/mac.c	2015-06-08 14:15:45.115688880 +0200
@@ -135,7 +135,7 @@
 	/* Setup the receive address */
 	e_dbg("Programming MAC Address into RAR[0]\n");
 
-	hw->mac.ops.rar_set(hw, hw->mac.addr, 0);
+	hw->mac.ops.rar_set(hw, hw->mac.perm_addr, 0);
 
 	/* Zero out the other (rar_entry_count - 1) receive addresses */
 	e_dbg("Clearing RAR[1-%u]\n", rar_count - 1);
Index: linux-3.19.8/drivers/net/ethernet/intel/e1000e/netdev.c
===================================================================
--- linux-3.19.8.orig/drivers/net/ethernet/intel/e1000e/netdev.c	2015-05-11 14:34:10.000000000 +0200
+++ linux-3.19.8/drivers/net/ethernet/intel/e1000e/netdev.c	2015-06-08 14:15:45.219020989 +0200
@@ -4516,7 +4516,8 @@
 	memcpy(netdev->dev_addr, addr->sa_data, netdev->addr_len);
 	memcpy(adapter->hw.mac.addr, addr->sa_data, netdev->addr_len);
 
-	hw->mac.ops.rar_set(&adapter->hw, adapter->hw.mac.addr, 0);
+	hw->mac.ops.rar_set(&adapter->hw, adapter->hw.mac.addr,
+			    adapter->hw.mac.rar_entry_count - 2);
 
 	if (adapter->flags & FLAG_RESET_OVERWRITES_LAA) {
 		/* activate the work around */
