Index: linux-3.19.8/drivers/input/misc/xen-kbdfront.c
===================================================================
--- linux-3.19.8.orig/drivers/input/misc/xen-kbdfront.c	2015-05-11 14:34:10.000000000 +0200
+++ linux-3.19.8/drivers/input/misc/xen-kbdfront.c	2015-06-05 15:24:11.175037693 +0200
@@ -168,8 +168,8 @@
 
 	if (abs) {
 		__set_bit(EV_ABS, ptr->evbit);
-		input_set_abs_params(ptr, ABS_X, 0, XENFB_WIDTH, 0, 0);
-		input_set_abs_params(ptr, ABS_Y, 0, XENFB_HEIGHT, 0, 0);
+		input_set_abs_params(ptr, ABS_X, 0, XENFB_TABLET_WIDTH, 0, 0);
+		input_set_abs_params(ptr, ABS_Y, 0, XENFB_TABLET_HEIGHT, 0, 0);
 	} else {
 		input_set_capability(ptr, EV_REL, REL_X);
 		input_set_capability(ptr, EV_REL, REL_Y);
Index: linux-3.19.8/include/xen/interface/io/fbif.h
===================================================================
--- linux-3.19.8.orig/include/xen/interface/io/fbif.h	2015-05-11 14:34:10.000000000 +0200
+++ linux-3.19.8/include/xen/interface/io/fbif.h	2015-06-05 15:24:11.275036498 +0200
@@ -137,6 +137,8 @@
 #ifdef __KERNEL__
 #define XENFB_WIDTH 800
 #define XENFB_HEIGHT 600
+#define XENFB_TABLET_WIDTH 32767
+#define XENFB_TABLET_HEIGHT 32767
 #define XENFB_DEPTH 32
 #endif
 
