Index: linux-3.19.8/drivers/char/tpm/tpm_tis.c
===================================================================
--- linux-3.19.8.orig/drivers/char/tpm/tpm_tis.c	2015-05-11 14:34:10.000000000 +0200
+++ linux-3.19.8/drivers/char/tpm/tpm_tis.c	2015-06-05 16:28:56.095344602 +0200
@@ -613,8 +613,8 @@
 	chip->vendor.manufacturer_id = vendor;
 
 	dev_info(dev,
-		 "1.2 TPM (device-id 0x%X, rev-id %d)\n",
-		 vendor >> 16, ioread8(chip->vendor.iobase + TPM_RID(0)));
+		 "1.2 TPM (vendor-id 0x%X device-id 0x%X, rev-id %d)\n",
+		 (vendor & 0xffff), vendor >> 16, ioread8(chip->vendor.iobase + TPM_RID(0)));
 
 	if (!itpm) {
 		probe = probe_itpm(chip);
