Index: linux-3.19.8/drivers/usb/xen-usbback/vusb.c
===================================================================
--- linux-3.19.8.orig/drivers/usb/xen-usbback/vusb.c	2015-06-09 15:45:22.649028795 +0200
+++ linux-3.19.8/drivers/usb/xen-usbback/vusb.c	2015-06-09 15:50:49.071911432 +0200
@@ -529,6 +529,23 @@
 	}
 }
 
+static int maybe_set_configuration(struct usb_device *dev, int configuration)
+{
+	struct usb_host_config *cp = NULL;
+        int i;
+
+        for (i = 0; i < dev->descriptor.bNumConfigurations; i++) {
+                if (dev->config[i].desc.bConfigurationValue ==
+                    configuration) {
+                        cp = &dev->config[i];
+                        break;
+                }
+        }
+        if (cp && cp == dev->actconfig)
+                return 0;
+        return dusb_set_configuration(dev, configuration);
+}
+
 static int setup_control_urb(struct vusb *vusb, usbif_request_t *req,
 				struct urb *urb)
 {
@@ -564,7 +581,7 @@
 			debug_print(LOG_LVL_DEBUG, "set config %d\n", confnum);
 
 			usb_lock_device(usbdev);
-			ret = dusb_set_configuration(usbdev, confnum);
+			ret = maybe_set_configuration(usbdev, confnum);
 			usb_unlock_device(usbdev);
 			if (ret == 0)
 				return (1);
