diff --git a/drivers/net/xen-netfront.c b/drivers/net/xen-netfront.c
index 6241bb1..2e1b15a 100644
--- a/drivers/net/xen-netfront.c
+++ b/drivers/net/xen-netfront.c
@@ -1974,15 +1974,26 @@ static void netback_changed(struct xenbus_device *dev,
 	dev_dbg(&dev->dev, "%s\n", xenbus_strstate(backend_state));
 
 	switch (backend_state) {
+	case XenbusStateUnknown:
+		/* if the backend vanishes from xenstore, close frontend */
+		if (!xenbus_exists(XBT_NIL, dev->otherend, "") &&
+		    (dev->state != XenbusStateClosed)) {
+			dev_warn(&dev->dev,
+				"backend vanished, closing frontend\n");
+			xenbus_frontend_closed(dev);
+			xennet_disconnect_backend(np);
+		}
+		break;
 	case XenbusStateInitialising:
 	case XenbusStateInitialised:
 	case XenbusStateReconfiguring:
 	case XenbusStateReconfigured:
-	case XenbusStateUnknown:
 		break;
 
 	case XenbusStateInitWait:
-		if (dev->state != XenbusStateInitialising)
+		/* allow reconnect if our state is either initialising, or closed */
+		if (dev->state != XenbusStateInitialising &&
+		    dev->state != XenbusStateClosed)
 			break;
 		if (xennet_connect(netdev) != 0)
 			break;
